### PR TITLE
To array functions

### PR DESCRIPTION
**Pull request by VinVaz**
In regard of issue #1616.
Some of the for..in loop were also changed to object methods, but there were situations that it was not clear if it was the best thing to do, since the object iterated, been an object variable referenced by "this", could have its value as undefined in some cases throwing an error when processed. So I have only changed a portion of them, I hope that's not a problem, but it's something to be aware of.

### CODE REVIEW

**Comment by ylemkimon** [comment]
`find()` and `findIndex()` are not supported by IE and old browsers, and they are not polyfilled as Babel cannot determine the type at compile time. I apologize I forgot to mention it in the issue.

File: contrib/auto-render/auto-render.js [hunk 1]
```diff
@@ -21,39 +20,38 @@ const renderMathInText = function(text, optionsCopy) {
...
-    }
+    });
 
     return fragment;
 };
 
 const renderElem = function(elem, optionsCopy) {
-    for (let i = 0; i < elem.childNodes.length; i++) {
-        const childNode = elem.childNodes[i];
+    elem.childNodes.forEach((childNode, i) => {
```
Comment (ylemkimon) [inline contrib/auto-render/auto-render.js]: `NodeList.prototype.forEach()` is not supported by IE and old browsers.

File: contrib/copy-tex/copy-tex.js [hunk 2]
```diff
@@ -12,10 +12,8 @@ document.addEventListener('copy', function(event) {
         return;  // default action OK if no .katex-mathml elements
     }
     // Preserve usual HTML copy/paste behavior.
-    const html = [];
-    for (let i = 0; i < fragment.childNodes.length; i++) {
-        html.push(fragment.childNodes[i].outerHTML);
-    }
+    const html = fragment.childNodes.map(node => node.outerHTML);
```
Comment (ylemkimon) [inline contrib/copy-tex/copy-tex.js]: `fragment.childNodes` is a `NodeList` and it doesn't have `map` method. You can use `Array.prototype.map.call`.

File: src/functions/verb.js [hunk 3]
```diff
@@ -21,17 +21,16 @@ defineFunction({
     },
     htmlBuilder(group, options) {
         const text = buildCommon.makeVerb(group, options);
-        const body = [];
         // \verb enters text mode and therefore is sized like \textstyle
         const newOptions = options.havingStyle(options.style.text());
-        for (let i = 0; i < text.length; i++) {
-            let c = text[i];
+        const body = text.map((element) => {
```
Comment (ylemkimon) [inline src/functions/verb.js]: `text` is a `String`, and it doesn't have `map` method.

File: src/Parser.js [hunk 4]
```diff
@@ -310,11 +310,10 @@ export default class Parser {
      */
     handleUnsupportedCmd(): AnyParseNode {
         const text = this.nextToken.text;
-        const textordArray = [];
 
-        for (let i = 0; i < text.length; i++) {
-            textordArray.push({type: "textord", mode: "text", text: text[i]});
-        }
+        const textordArray = text.map((element) => {
```
Comment (kevinbarabash) [inline src/Parser.js]: This loop can't be converted b/c `text` is a `string`.

File: src/buildCommon.js [hunk 5]
```diff
@@ -276,11 +276,10 @@ const makeOrd = function<NODETYPE: "spacing" | "mathord" | "textord">(
         } else if (ligatures.hasOwnProperty(text) &&
                    fontName.substr(0, 10) === "Typewriter") {
             // Deconstruct ligatures in monospace fonts (\texttt, \tt).
-            const parts = [];
-            for (let i = 0; i < text.length; i++) {
-                parts.push(makeSymbol(text[i], fontName, mode, options,
-                                      classes.concat(fontClasses)));
-            }
+            const parts = text.map((element) => {
```
Comment (kevinbarabash) [inline src/buildCommon.js]: This loop can't be converted b/c `text` is a `string`.

File: src/domTree.js [hunk 6]
```diff
@@ -60,24 +60,20 @@ const toNode = function(tagName: string): HTMLElement {
     node.className = createClass(this.classes);
 
     // Apply inline styles
-    for (const style in this.style) {
-        if (this.style.hasOwnProperty(style)) {
-            // $FlowFixMe Flow doesn't seem to understand span.style's type.
-            node.style[style] = this.style[style];
-        }
-    }
+    Object.keys(this.style).map((style) => {
```
Comment (kevinbarabash) [inline src/domTree.js]: Please use `Object["key"]` to avoid babel polyfilling this for us.

File: src/domTree.js [hunk 7]
```diff
@@ -60,24 +60,20 @@ const toNode = function(tagName: string): HTMLElement {
...
+        node.style[style] = this.style[style];
+    });
 
     // Apply attributes
-    for (const attr in this.attributes) {
-        if (this.attributes.hasOwnProperty(attr)) {
-            node.setAttribute(attr, this.attributes[attr]);
-        }
-    }
+    Object.keys(this.attributes).map((attr) => {
```
Comment (kevinbarabash) [inline src/domTree.js]: Please use `Object["key"]` to avoid babel polyfilling this for us.

File: src/functions/verb.js [hunk 8]
```diff
@@ -21,7 +21,6 @@ defineFunction({
     },
     htmlBuilder(group, options) {
         const text = buildCommon.makeVerb(group, options);
-        const body = [];
```
Comment (kevinbarabash) [inline src/functions/verb.js]: Why was this removed?

**Comment by VinVaz** [comment]
I can see there is still an error, I'm working on it

**Comment by VinVaz** [comment]
Since my first PR every time I send a push to my fork it seems it updates the PR for you, sorry for the noise

**Comment by kevinbarabash** [comment]
@VinVaz no worries.  You should be able to run `yarn test` locally which should help with identify some of the issues before pushing.

**Comment by VinVaz** [comment]
The local test only throws a Spawn Error and I have not been able to find its source.

**Comment by kevinbarabash** [comment]
My guess is that there's something wrong with your dev environment.  I'm guessing you're on Windows.  In order to narrow things down, I would try running each command separately that the `test` script from package.json runs and then Google for the failure in whichever one fails.  In the meantime, circleci has reported the following test failures:
```
Error ---------------------------------------------------------------------------------- src/environments/array.js:244:9

Cannot assign `outrow` to `body[r]` because:
 - an indexer property is missing in `span` [1].
 - an indexer property is missing in `span` [2].

   src/environments/array.js:244:9
   244|         body[r] = outrow;
                ^^^^^^^

References:
   src/buildCommon.js:348:4
   348| ): DomSpan {
           ^^^^^^^ [1]
   src/buildCommon.js:523:67
   523| const makeVList = function(params: VListParam, options: Options): DomSpan {
                                                                          ^^^^^^^ [2]
```

**Comment by kevinbarabash** [comment]
Also, something must not have worked with `CRLF to LF` since there's still 255 changes files according to GitHub.

**Comment by ylemkimon** [comment]
The best way to fix these problems would be to save the work elsewhere, delete the whole project, and cloning a fresh copy. (https://xkcd.com/1597/ 😄)

**Comment by kevinbarabash** [comment]
Closing due to lack of activity.  @VinVaz feel free to reopen a fresh PR if you have time.
